ENH: Centralize fMRIPrep's and MRIQC's guidelines for Docker & DataLad (#47)
nipreps/nipreps.github.io: docs/apps/datalad.md

## *DataLad* and *Docker*

> 

Reviewer: could/should I propose (separate PR on top?) in this section and/or https://github.com/nipreps/nipreps.github.io/blob/HEAD/docs/apps/singularity.md file to mention our https://github.com/ReproNim/containers which contains (automatically updates) pre-created singularity images for all bids-apps (thus including fmriprep, mriqc), and providing some helpers to streamline their use and more guaranteed reproducibility (isolated environment execution etc).

Note that wrapper also tries to support non-Linux systems (OSX) where we could run singularity under docker. Or could also be used on Linux if there is no singularity installation.

https://github.com/ReproNim/containers?tab=readme-ov-file#runnable-script provides a "typical" use example based on mriqc.  

https://github.com/OpenNeuroDerivatives/ by @jbwexler (and @effigies ?) use that ReproNim/containers as a subdatset  archive of the images with `reproman run` for execution.

Reviewer: also that should include YODA aspects whenever talking about containers... with them it becomes possible to encapsulate all digital objects nicely and reproducibly (there is no guarantee that docker:// would later give you the images used etc)

Author: Yes! Mentioning ReproNim on a separate PR would be fantastic.

Apps may be able to identify if the input dataset is handled with

> *DataLad* or *Git-Annex*, and pull down linked data that has not

Reviewer: ```suggestion
[*DataLad*](https://www.datalad.org/) or [*git-annex*](https://git-annex.branchable.com), and pull down linked data that has not
```
to use correct (absent) capitalization in git-annex and also direct users if desired to projects' websites.

> been fetched yet.

Reviewer: If user knows that data is fetched already, or data is to be fetched by an external execution wrapper (e.g. `datalad containers-run` or `datalad run` or `reproman run` used by openneuro), any problem with using datalad/git-annex could then be avoided by utilizing `--no-datalad-get` for mriqc.  It was also our with @asmacdo workaround for the 

- https://github.com/nipreps/mriqc/issues/1307

That is why we even added it as a default to our repronim/containers collection configuration for `datalad containers-run`: https://github.com/ReproNim/containers/pull/133/files  

I think it might be worth noting somewhere here in these docs (do not see mentioning of `--no-datalad-get` anywhere in this repo ATM).

Author: Thanks, I think more info about this would be important. I'll add a terse note here and a more complete description in MRIQC's documentation.

Author: Added in 44360cb. Please send further suggestions or reopen this conversation if that fell short of addressing the issue.

* the uid who is *executing MRIQC* within the container must
  have sufficient permissions to write in the tree.

!!! tip "Check *ReproNim* if the suggestions here did not work"

>     The actions suggested here are relatively fundamental, and your
>     settings may showcase specific circumstances that render them
>     insufficient.

Reviewer 2: ```suggestion
    The actions suggested here are expected to work in most circumstances,
    but your system may have specific circumstances that require additional
    or alternative approaches.
```

>     For a more detailed discussion about orchestrating containers, version
>     control, and reproducible pipelines, please check
>     [the *ReproNim* guidelines](https://www.repronim.org/).

Reviewer: ```suggestion
    For instance,  [the *ReproNim* project](https://www.repronim.org/) maintains 
    [ReproNim/containers](https://github.com/ReproNim/containers) -- a
    DataLad dataset with ready-to-use Singularity images for released BIDS-apps, NeuroDesktop applications,
    and other containers. 
    Its [README.md](https://github.com/ReproNim/containers?tab=readme-ov-file#runnable-script) guides through an approach via that dataset with *built-in* execution helper taking care about bind-mounts,
    proxying critical git configuration, and potentially executing singularity images via Docker (e.g., under OSX).
```

git-annex: automatic initialization failed due to above problems']
```

Confusingly, following the suggestion from *DataLad*
(just propagated from *Git*) of executing
`git config --global --add safe.directory /data` will not work in this
case, because this line must be executed within the container.
However, containers are *transient* and the setting this configuration
on *Git* will not be propagated between executions unless advanced

> actions are taken (such as mounting a *Home* folder with the necessary settings).

Reviewer: ```suggestion
actions are taken (such as mounting a *HOME* folder with the necessary settings).
```

IMHO more "appropriate" way to refer to it.

FWIW, that is a great point and for aforementioned repronim/containers "singularity helper" we indeed populate gitconfig with such settings from the outside user environment: https://github.com/ReproNim/containers/blob/master/scripts/singularity_cmd#L78

In general, to make troubleshooting within container "space" is easier on users' mind, and also to allow for various other, possibly suboptimal practices such as "full paths in the scripts" naive users often do, and with YODA in mind, I tend to just recommend to

- bind `$PWD` as `$PWD` within container space
- ensure that all necessary components are reachable from under `$PWD/` (YODA).

This way 
- easy to "transfer" configurations from "outside" inside container execution space
- there is a clear parallel/easier to compare "outside" and "inside" executions since paths would be the same
- with YODA in mind -- better self-containment etc...

With all that in mind -- I wonder if similar practice (`$PWD:$PWD` bind) could be recommended here?

Related aspect then also would be `workdir` specification and its addition to `.gitignore` if also residing under the same folder.

Author: I'm happy to add these recommendations, but also happy to make an admonition tip where these possibilities are mentioned and a link to ReproNim for the best description.

Author: Added in 44360cb. Please send further suggestions and/or reopen this conversation if that fell short of addressing the issue.
